gnomad_qc/v4/annotations/generate_freq_genomes.py

```diff
@@ -1168,16 +1218,310 @@ def get_age_distribution(
     return age_expr
 
 
+def get_pop_diff_v3_vds_group_membership(
+    v3_vds: hl.vds.VariantDataset,
+    meta_ht: hl.Table,
+) -> hl.Table:
+    """
+    Get the group membership for samples in pop_diff for the v3.1 VDS.
+
+    :param v3_vds: VDS with the v3.1 release samples.
+    :param meta_ht: Table with the updated HGDP + 1KG sample metadata.
+    :return: Table with the group membership for samples in pop_diff.
+    """
+    v3_sites_meta_ht = v3_vds.variant_data.cols()
+    v3_sites_meta_ht = v3_sites_meta_ht.filter(v3_sites_meta_ht.meta.release)
+
+    logger.info(
+        "Annotating call stats group membership for pop diff release samples..."
+    )
+    # Need to make sure we are using the names that are in the vds for the release
+    # samples.
+    # Note: the samples in the pop_diff HT are samples in the to-be-split 'Han' and
+    # 'Papuan' populations AND their 'gnomad_release' status hasn't changed.
+    v3_sites_meta_rename_ht = v3_sites_meta_ht.key_by(
+        s_no_prefix=v3_sites_meta_ht.s.replace("v3.1::", "")
+    )
+    pop_diff_sample_ht = get_updated_release_samples(meta_ht)[0]
+    pop_diff_sample_ht = pop_diff_sample_ht.key_by(
+        s=v3_sites_meta_rename_ht[pop_diff_sample_ht.s].s
+    )
+    pop_diff_group_membership_ht = get_group_membership_ht_for_an(
+        pop_diff_sample_ht, only_pop_diff=True
+    )
+    pop_diff_group_membership_ht = pop_diff_group_membership_ht.checkpoint(
+        new_temp_file("group_membership_pop_diff", "ht")
+    )
+
+    return pop_diff_group_membership_ht
+
+
+def patch_v4_genomes_callstats(freq_ht: hl.Table) -> hl.Table:
+    """
+    Patch the call stats for some inconsistent variant calls found during validity checks.
+
+    14 variants were found to have inconsistent calls between v3.1 and v4.0 genomes
+    release, we determined the reason for each of these by manual inspection and
+    this function applies the needed patches to these variants.
+
+    .. note::
+        This is a temporary fix until we can recompute the call stats for the all the
+        samples.
+
+    :param freq_ht: Table with the call stats for the v4.0 genomes release.
+    :return: Table with the patched call stats for the v4.0 genomes release.
+    """
+    # For most of the variants, there is a single Han sample that was 0/1 in v3.1, but
+    # from the dense MT in v4.0 frequencies it was given a missing genotype. These
+    # variants are not in v3, which is why they get a missing genotype in the v4.0
+    # freq. We don’t understand why they were 0/1 in v3.1, we would expect them to
+    # have been missing. It should be 0/1 if missing v3 frequency had been handled
+    # correctly in the fix for high AB hets.
+    freq_groups1 = hl.set(
```

Comment:
I know we're short of time, it took me some time to understand why there are so many groups, now I understand it's a combination of problem, sample's sex and sample's population. 
Since patch1_expr ~ patch4_expr are identical, can we make a small function like this? 
```commandline
def patch_add(freq_groups: hl.set_expr, freq: hl.expr) -> hl.expr:
    return hl.map(
        lambda x, m: hl.if_else(
            freq_groups.contains(m),
            hl.struct(
                AC=x.AC + 1,
                AF=(x.AC + 1) / (x.AN + 2),
                AN=x.AN + 2,
                homozygote_count=x.homozygote_count,
            ),
            x,
        ),
        freq_ht.freq,
        freq_ht.freq_meta,
    )
```

Reply:
not at this time please

Reply:
we can refactor after

Reply:
okay, sounds good.